### PR TITLE
Spatial white Gaussian noise

### PR DESCRIPTION
**Pull request by brendankeith**
This PR provides a new `LinearFormIntegrator` for spatial white Gaussian noise.

The target problem is the linear SPDE `a(u,v) = F(v)` with `F(v) := <Ẇ,v>`, where `Ẇ` is spatial white Gaussian noise. When the Galerkin method is used to discretize this problem into a linear system of equations `Ax = b`, the RHS is a Gaussian random vector `b~N(0,M)` whose covariance matrix is the same as the mass matrix `M_ij = (v_i,v_j)`. This property can be ensured if `b = H w`, where `HHᵀ = M` and each component `w_i~N(0,1)`.

There is much flexibility in how we may wish to define `H`. In this PR, we follow [Section 4.1, 1] and define `H = Pᵀ diag(L_e)`, where `P` is the global-to-local dof assembly matrix and `diag(L_e)` is a block-diagonal matrix with `L_e L_eᵀ = M_e`, where `M_e` is the element mass matrix for element `e`. A straightforward computation shows that `HHᵀ = Pᵀ diag(M_e) P = M`, as necessary.

The unit tests in this PR use the Monte Carlo method to verify`E[b] = 0` and `E[bbᵀ] = M`.

### NOTES
- To experiment with this PR, change `line 173` of `ex1.cpp` to `b.AddDomainIntegrator(new WhiteGaussianNoiseDomainLFIntegrator());`
- This PR will be merged with https://github.com/mfem/mfem/pull/2832 for a forthcoming MFEM example on the generation of Matérn random fields

### REFERENCES
[1] Croci, M., Giles, M. B., Rognes, M. E., & Farrell, P. E. (2018). Efficient white noise sampling and coupling for multilevel Monte Carlo with nonnested meshes. SIAM/ASA Journal on Uncertainty Quantification, 6(4), 1630-1655.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2868](https://github.com/mfem/mfem/pull/2868) | @brendankeith | @tzanio | @bslazarov + @jwang125 + @jamiebramwell | 3/20/22 | 4/11/22 | 4/17/22 | |

<details>
<summary>PR Checklist</summary>
    
- [x] Code builds.
- [x] Code passes `make style`.
- [x] Update `CHANGELOG`:
    - [x] Is this a new feature users need to be aware of? New or updated example or miniapp?
- [x] Update `INSTALL`:
    - [x] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [x] Have the version ranges for any required or optional libraries changed?
- [x] New capability:
   - [x] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [x] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [x] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [x] Consider adding new sample runs in existing examples to highlight the new capability.
   - [x] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
- [x] Run `make unittest` to make sure all unit tests pass.
- [x] Run the tests in `tests/scripts`.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
See the PR checklist in the [description](https://github.com/mfem/mfem/pull/2868#issue-1157901500)

File: fem/lininteg.hpp [hunk 1]
```diff
@@ -580,6 +582,105 @@ class QuadratureLFIntegrator : public LinearFormIntegrator
...
+public:
+
+   /** @brief Sets the @a seed_ of the random number generator */
+#ifdef MFEM_USE_MPI
+   WhiteGaussianNoiseDomainLFIntegrator(int seed_ = 0)
+      : LinearFormIntegrator(), comm(MPI_COMM_NULL)
+   {
+      if (seed_ > 0) { SetSeed(seed_); }
+   }
+
```
Comment (bslazarov) [inline fem/lininteg.hpp:624]: add documentation for the constructor - also specify how the seed is used in the initialization

File: fem/lininteg.hpp [hunk 2]
```diff
@@ -580,6 +582,105 @@ class QuadratureLFIntegrator : public LinearFormIntegrator
...
+   void SetSeed(int seed)
+   {
+      generator.seed(seed);
+   }
+
+   using LinearFormIntegrator::AssembleRHSElementVect;
+   virtual void AssembleRHSElementVect(const FiniteElement &el,
+                                       ElementTransformation &Tr,
+                                       Vector &elvect);
+   void SaveFactors(int NE)
```
Comment (bslazarov) [inline fem/lininteg.hpp:651]: Add documentation for SaveFactors

File: fem/lininteg.hpp [hunk 3]
```diff
@@ -580,6 +582,105 @@ class QuadratureLFIntegrator : public LinearFormIntegrator
...
+   {
+      save_factors = true;
+      ResetFactors(NE);
+   }
+
+   void ResetFactors(int NE)
```
Comment (bslazarov) [inline fem/lininteg.hpp:657]: Add documentation for ResetFactors

File: linalg/densemat.hpp [hunk 4]
```diff
@@ -524,12 +524,49 @@ void AddMult_a_VWt(const double a, const Vector &v, const Vector &w,
...
+/** Abstract class that can compute factorization of external data and perform various
     operations with the factored data. */
-class LUFactors
+class Factors
 {
 public:
+
    double *data;
+
+   Factors() { }
```
Comment (bslazarov) [inline linalg/densemat.hpp:535]: Documentation?

File: linalg/densemat.hpp [hunk 5]
```diff
@@ -524,12 +524,49 @@ void AddMult_a_VWt(const double a, const Vector &v, const Vector &w,
...
+
+   Factors(double *data_) : data(data_) { }
```
Comment (bslazarov) [inline linalg/densemat.hpp:537]: Documentation?

File: linalg/densemat.hpp [hunk 6]
```diff
@@ -524,12 +524,49 @@ void AddMult_a_VWt(const double a, const Vector &v, const Vector &w,
...
+
+   virtual bool Factor(int m, double TOL = 0.0)
```
Comment (bslazarov) [inline linalg/densemat.hpp:539]: Documentation?

File: linalg/densemat.hpp [hunk 7]
```diff
@@ -524,12 +524,49 @@ void AddMult_a_VWt(const double a, const Vector &v, const Vector &w,
...
+   {
+      mfem_error("Factors::Factors(...)");
+      return false;
+   }
+
+   virtual double Det(int m) const
```
Comment (bslazarov) [inline linalg/densemat.hpp:545]: Documentation

File: linalg/densemat.hpp [hunk 8]
```diff
@@ -524,12 +524,49 @@ void AddMult_a_VWt(const double a, const Vector &v, const Vector &w,
...
+   {
+      mfem_error("Factors::Det(...)");
+      return 0.;
+   }
+
+   virtual void Solve(int m, int n, double *X) const
```
Comment (bslazarov) [inline linalg/densemat.hpp:551]: Documentation

File: linalg/densemat.hpp [hunk 9]
```diff
@@ -524,12 +524,49 @@ void AddMult_a_VWt(const double a, const Vector &v, const Vector &w,
...
+   {
+      mfem_error("Factors::Solve(...)");
+   }
+
+   virtual void GetInverseMatrix(int m, double *X) const
```
Comment (bslazarov) [inline linalg/densemat.hpp:556]: Documentation

File: linalg/densemat.hpp [hunk 10]
```diff
@@ -630,24 +667,87 @@ class LUFactors
...
+    SPD matrix and perform various operations with the factored data. */
+class CholeskyFactors : public Factors
+{
+public:
+
+   /** With this constructor, the (public) data should be set
+       explicitly before calling class methods. */
+   CholeskyFactors() : Factors() { }
+
+   CholeskyFactors(double *data_) : Factors(data_) { }
```
Comment (bslazarov) [inline linalg/densemat.hpp:680]: Documentation

File: linalg/densemat.hpp [hunk 11]
```diff
@@ -630,24 +667,87 @@ class LUFactors
...
-   LUFactors lu;
+   Factors * factors = nullptr;
+   bool spd = false;
 
+   void Init(int m);
+   bool own_data = false;
 public:
    /// Default constructor.
-   DenseMatrixInverse() : a(NULL), lu(NULL, NULL) { }
+   DenseMatrixInverse(bool spd_=false) : a(NULL), spd(spd_) { Init(0); }
```
Comment (bslazarov) [inline linalg/densemat.hpp:743]: Documentation?

**Review by bslazarov — APPROVED** [review]

File: linalg/densemat.hpp (hunk 4 above)
Comment (brendankeith) [reply linalg/densemat.hpp:535]: I will wait to make changes in this file until https://github.com/mfem/mfem/pull/2887 is merged.

**Comment by jwang125** [comment]
Compiled with no problem with camke. However, tThe testing file is not added to CMakeList.txt in tests/ . Also has no output it seems. I did add the test and some output and can confirm that it runs smoothly.

**Review by jwang125 — APPROVED** [review]
Other than my comments about test_white_noise.cpp, everything works well for me.

**Comment by tzanio** [comment]
Thank you @jwang125.

> However, the testing file is not added to CMakeList.txt in tests/ . Also has no output it seems.

@brendankeith, can you address this ☝️ ?

**Comment by brendankeith** [comment]
> Thank you @jwang125.
> 
> > However, the testing file is not added to CMakeList.txt in tests/ . Also has no output it seems.
> 
> @brendankeith, can you address this ☝️ ?

Of course!

**Comment by brendankeith** [comment]
@tzanio Is there anything left to do here before merging into next?

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: tests/unit/fem/test_white_noise.cpp [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+      ParBilinearForm a(&fespace);
+      a.AddDomainIntegrator(new MassIntegrator());
+      a.Assemble();
+
+      HypreParMatrix M;
+      Array<int> empty;
+      a.FormSystemMatrix(empty,M);
+
+      // Compare population covariance to mass matrix
+      HypreParMatrix * diff = Add(1., *ParC, -1., M);
```
Comment (v-dobrev) [inline tests/unit/fem/test_white_noise.cpp:209]: This line causes failures in our nightly tests. The issue is that (at least with older hypre versions) `Add` requires that the two `HypreParMatrix`es have the same `col_map_offd` arrays which in this case is not true.

The alternative that can be used here is:
```suggestion
      M *= -1.;                                                                                                                     
      HypreParMatrix * diff = ParAdd(ParC, &M);                                                                                     
```

Comment (v-dobrev) [inline tests/unit/fem/test_white_noise.cpp:209]: This line causes failures in our nightly tests. The issue is that (at least with older hypre versions) `Add` requires that the two `HypreParMatrix`es have the same `col_map_offd` arrays which in this case is not true.

The alternative that can be used here is:
```suggestion
      M *= -1.;                                                                                                                     
      HypreParMatrix * diff = ParAdd(ParC, &M);                                                                                     
```

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Review by tzanio — COMMENTED** [review]
Thanks for the contribution @brendankeith 👍 

@bslazarov, @jwang125, @jamiebramwell and @v-dobrev -- thank you for reviewing!

**Comment by brendankeith** [comment]
> Thanks for the contribution @brendankeith 👍
> 
> @bslazarov, @jwang125, @jamiebramwell and @v-dobrev -- thank you for reviewing!

Also many many many thanks to @psocratis for helping me with this!!!

**Pull request merged**